DOCSP-50185: Connection troubleshooting (#272)
mongodb/docs-php-library: source/connect/connection-troubleshooting.txt

.. meta::
   :keywords: code example, disconnected, deployment

This page offers potential solutions to issues that you might encounter
when using the {+library-short+} to connect to a MongoDB deployment.

.. note::

   This page addresses only connection issues. If you encounter other
   issues when using MongoDB or the {+library-short+}, visit the following resources:

   - The :ref:`Issues & Help <php-issues-and-help>` page for
     information about reporting bugs, contributing to the library, and
     finding more resources
   - The :community-forum:`MongoDB Community Forums </tag/php>` for
     questions, discussions, or general technical support

Server Connection Errors
------------------------

When an issue occurs when you attempt to connect to a server, the {+driver-short+}
returns an error message. If this error resembles the following message, it
indicates that the library cannot connect to a MongoDB deployment:

.. code-block:: none
   :copyable: false

   No suitable servers found (`serverSelectionTryOnce` set):
   [connection refused calling hello on 'localhost:27017']

The following sections describe methods that might help resolve the issue.

> Check the Connection String

Reviewer: Thoughts about changing all occurrences of "connection string" to "connection URI"? Seems to be the wording we've been trending towards across our docs.

~~~~~~~~~~~~~~~~~~~~~~~~~~~

Verify that the hostname and port number in the connection string are both
accurate. In the sample error message, the hostname is ``127.0.0.1`` and the

> port is ``27017``. The default port value for an instance of MongoDB Server is
> ``27017``, but you can configure MongoDB to listen on another port.

Reviewer: ```suggestion
port is ``27017``. The default port value for a {+mdb-server+} deployment is
``27017``, but you can configure MongoDB to listen on another port.
```

When connecting to a replica set, include all the replica set hosts in
your connection string. Separate each of the hosts in the connection
string with a comma. This enables the library to establish a connection if
one of the hosts is unreachable.

To learn how to specify multiple replica set hosts, see the
:ref:`Replica Sets <php-connection-replica-set>` section of the
Choose a Connection Target guide.

Configure the Firewall
~~~~~~~~~~~~~~~~~~~~~~

If your MongoDB deployment is hosted behind a firewall, ensure the port
on which MongoDB listens is open in the firewall. If your deployment
listens on the default network port, ensure that port ``27017`` is
open in the firewall. If your deployment listens on a different port,
ensure that port is open on the firewall.

.. warning::

   Do not open a firewall port unless you are sure that it is the one
   that your MongoDB deployment listens on.

Authentication Errors
---------------------

> The {+driver-short+} may be unable connect to a MongoDB deployment if
> the authorization is not configured correctly. In these cases, the library

Reviewer: ```suggestion
The {+driver-short+} may fail to connect to a MongoDB deployment if
the authorization is not configured correctly. In these cases, the library
```

Author: I'll keep as "unable" for [style guide](https://www.mongodb.com/docs/meta/style-guide/writing/use-positive-statements/#avoid-negative-words) reasons

raises an error message similar to the following message:

.. code-block:: none
   :copyable: false

   Authentication failed.

The following sections describe methods that may help resolve the issue.

Check the Credentials Formatting
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

One of the most common causes of authentication issues is invalid
credentials formatting in the MongoDB connection string.

.. tip::

   To learn more information about using connection strings,
   see :ref:`Connection URI <php-connection-uri>` in the
   Create a MongoDB Client guide.

If your connection string contains a username and password, ensure that
they are correctly formatted.

.. note::

>    If your username or password includes any of the following characters, you
>    must `percent-encode <https://tools.ietf.org/html/rfc3986#section-2.1>`__ it:

Reviewer: ```suggestion
   You
   must `percent-encode <https://tools.ietf.org/html/rfc3986#section-2.1>`__ the following characters if they appear in your username or password:
```

in an options parameter:

.. code-block:: php
   :copyable: false

   $uriOptions = [
      'username' => '<username>',
      'password' => '<password>',
      'authSource' => '<authentication database>',
      'authMechanism' => 'SCRAM-SHA-1',
   ];

   $client = new MongoDB\Client(
      'mongodb://<hostname>:<port>',
      $uriOptions,
   );

To learn more about specifying authentication mechanisms, see the :ref:`php-auth`
section.

Verify User Is in Authentication Database
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

When using a username and password-based authentication method,
the username must be defined in the authentication database.

The default authentication database is the ``admin`` database.
To use a different database for authentication, specify the
``authSource`` option in the connection string.

The following example instructs MongoDB to use the ``users`` database
as the authentication database:

.. code-block:: php
   :copyable: false

   $uri = 'mongodb://<username>:<password>@<hostname>:<port>/?authSource=users';
   $client = new MongoDB\Client($uri);

DNS Resolution Errors
---------------------

> The {+driver-short+} may be unable to resolve your DNS connection. When this

Reviewer: ```suggestion
The {+driver-short+} may fail to resolve your DNS connection. When this
```
